### PR TITLE
contributing-guidelines-proposal

### PR DESCRIPTION
**Pull request by gkieszkowski**
| | |
|:---|:---|
|**_Originating Project/Creator_**| |
|**_Affected Component_**|  |
|**_Affected Architectures(s)_**|  |
|**_Related Issue(s)_**|  |
|**_Has Unit Tests (y/n)_**|  |
|**_Builds Without Errors (y/n)_**|  |
|**_Unit Tests Pass (y/n)_**|  |
|**_Documentation Included (y/n)_**|  |

---
## Change Description
added contributing.md file as a proposal 

## Rationale
Helping out new contributors
## Testing/Review Recommendations
no need

## Future Work
We need help from maintainers and contributors to edit this file before PR will be merged

### CODE REVIEW

**Comment by gkieszkowski** [comment]
@LeStarch I would appreciate if you could help with this file by adding branch naming convention and commit message convention as you are maintainer and now is a good time to create some standard.
Also if anyone wants to edit or add some lines feel free as this is just a proposal.

**Comment by astroesteban** [comment]
The [NASA cFS](https://github.com/nasa/cFS/blob/main/CONTRIBUTING.md) repo has a pretty good CONTRIBUTING.md file we could use as a reference. Another good example, albeit more verbose, is the [Atom](https://github.com/atom/atom/blob/master/CONTRIBUTING.md) CONTRIBUTING.md file. There's a lot of great content in that one. I also really like [VS Code's](https://github.com/microsoft/vscode/blob/main/CONTRIBUTING.md) one.

I can contribute more to this on the weekend.

Thanks for getting this going @gkieszkowski!

File: CONTRIBUTING.md
Comment (LeStarch) [inline CONTRIBUTING.md:28]: When doing a fork and PR model, do we need a branch naming convention?  Is there a typical recommendation here?

**Comment by LeStarch** [comment]
Yes, I will echo thanks for getting this in the mix!  I added one comment for now, and I need to read the above mentioned CFS link to see a good example.  Then I will add more feedback.

File: CONTRIBUTING.md
Comment (astroesteban) [reply CONTRIBUTING.md:28]: I typically do:
- feat/a-new-feature-branch
- fix/a-bugfix-branch
- hotfix/emergency-fix-branch
- release/a-release-branch

You _can_ get more granular with names like:
- docs/branch-only-changes-docs
- ci/changes-to-ci-or-tooling

But I feel like it gets more in the way than actually helping.

Comment (LeStarch) [reply CONTRIBUTING.md:28]: @astroesteban I use a similar naming scheme too...but since these branches will live on other people's forks, do we really need to tell them how to name their branches?  The only way that branch has into F´ is through a PR, which will put it in an already established branch.

We certainly can specify a naming pattern for branches, but shoul I reject a PR with a bad branch name that is outside the repo?  I likely would not reject it, and thus I am wondering if this is needed at all.

It certainly could be the case that that we want it to help new users choose branch names even if we won't strictly enforce it. If that is the reason to include it then I am on board!

Thoughts?

Comment (astroesteban) [reply CONTRIBUTING.md:28]: I agree with your points. I think the naming convention makes more sense when working directly on the main repo.

 Given that the pull request is the entry point to the main F' repo for contributors it doesn't really make sense to enforce a branching schema for contributors since what they do is on _their_ fork and shouldn't be dictated by the upstream F' repo since contributors "own" their fork.

Additionally, no other popular open-source project CONTRIBUTING.md I've looked at (OpenCV, TensorFlow, PyTorch, NumPy, etc) dictates a branching schema.

Comment (gkieszkowski) [reply CONTRIBUTING.md:28]: I guess you might be right about branch naming it may be an overkill and should not be in contributing guidelines.
We need to remove it than. 
In general I do feel that we could really look at contributing guidelines mentioned above and I do not mind totally changing my draft as compared to atom it is not as good.
Guess the key for creating those guidelines is answering following questions:
1. what information would help newcomers easily get involved in nasa/fprime?
2. what information would impact contributors behaviour in such way that would help maintainers with project management? 

Probably too much information or information which does not bring value to contributor's or maintainer's work could be overkill and create confusion (for ex. - branch naming).

**Comment by gkieszkowski** [comment]
Hi,
I have edited draft a bit in order to push it further. It needs a lot of work but I hope someone will pick it up later on and probably I will, as soon as I will know more good practices in nasa/fprime community.
Converting draft into PR.

**Comment by LeStarch** [comment]
Do you mind if I edit on this PR to suggest some things?

**Comment by LeStarch** [comment]
Some other questions:

1. Should links point to the published documentation or the internal files that build into that documentation?
2. Could you elaborate on "The second is to go with "start finishing, stop starting" principle from pull system."?  I don't quite follow what you mean there?

**Comment by gkieszkowski** [comment]
> Do you mind if I edit on this PR to suggest some things? I guess I should elaborate: I like this start, and long term I'd like us to get to the verbosity of the CFS contribution guidelines....but for now, I'd like to edit on this PR to fill in some much needed information specific to F´.

I dont mind it and really hope you will do that! Would appreciate it so much.

**Comment by gkieszkowski** [comment]
> Some other questions:
> 
> 1. Should links point to the published documentation or the internal files that build into that documentation?
> 2. Could you elaborate on "The second is to go with "start finishing, stop starting" principle from pull system."?  I don't quite follow what you mean there?

1. It is totally up to you. What you think is more beneficial?
2. The principle helps direct focus to already open PR, issues etc. For example if you have already open PRs waiting for review the longer PRs are waiting for merge and more the work on them is done the bigger probability that there will be merge conflicts and more work. In most cases many meaningful contributions are driven and done by core contributors and maintainers. So for example lets say you have 3 open PR which are crucial for project and some new contributor comes in. He just drives through codebase and got some improvements but instead checking first open issues or open PRs he starts new work. So if contriburing guidelines ask newcomers to checkout open PRs and issues found by core members and maintainers we could possibly direct people to topics, issues and work that has been already in progress.

**Comment by LeStarch** [comment]
@gkieszkowski I finally edited this with my thoughts.  Want to take a look and let me know what you think?

File: CONTRIBUTING.md
Comment (Joshua-Anderson) [inline CONTRIBUTING.md:44]: You might want to clarify why core framework changes are reviewed carefully

**Review by Joshua-Anderson — APPROVED** [review]
Look great!

File: CONTRIBUTING.md
Comment (LeStarch) [reply CONTRIBUTING.md]: Fixed

**Pull request merged**
